embedded-workshop-book/src/uarte-implementation.md

```diff
@@ -0,0 +1,183 @@
+use hal::uarte;
+```
+
+###  Step 3: Add `struct Uarte`
+
+✅ Add `struct Uarte` that serves as a wrapper for the `UARTE1` instance.
+The struct has one field labelled `inner`, it contains the `UARTE1` instance: `hal::Uarte<hal::pac::UARTE1>`.
+
+<details>
+  <summary>Solution</summary>
+
+```rust
+pub struct Uarte {
+    inner: hal::Uarte<hal::pac::UARTE1>,
+}
+```
+</details>
+
+### Step 4: Bring up the peripheral in the `fn init()`
+
+✅ Take a closer look at the definition of the `uarte::Pins` struct in the `nrf-hal`. Compare the pin type configurations with the ones you have already imported in `lib.rs`. Add the ones you're missing. 
+
+✅ Create an instance of this struct in `fn init()` with the appropriate pins and configurations. Set the output pin's level to `Level::High`.
+Note, that the third and fourth pin are each wrapped in an `Option`. 
+
+✅ Create an interface to the UARTE1 instance with `uarte::Uarte::new(...)` that you bind to a variable. This instantiating method takes four arguments:
+* The `UARTE1` instance can be found in the `periph` variable.
+* Your instance of the `uarte::Pins` struct.
+* Set parity to `Parity::INCLUDED` 
+* set the baud rate to `Baudrate::BAUD115200`.
+
+<details>
+  <summary>Solution</summary>
+
+```rust
+  let pins =  hal::uarte::Pins {
+            rxd: pins.p0_08.degrade().into_floating_input(),
+            txd: pins.p0_06.degrade().into_push_pull_output(Level::High),
+            cts: Some(pins.p0_07.degrade().into_floating_input()),
+            rts: Some(pins.p0_05.degrade().into_push_pull_output(Level::High)),
+        };
+       
+
+        let uarte = hal::uarte::Uarte::new(periph.UARTE1, pins, Parity::INCLUDED, Baudrate::BAUD115200);
+```
+</details>
+
+### Step 5: Board struct
+
+✅  Add a field for the `Uarte` struct in the `Board` struct. 
+add the field to the instance of the `Board` struct in `fn init()`.
+
+<details>
+  <summary>Solution</summary>
+
+```rust
+
+pub struct Board {
+    /// LEDs
+    pub leds: Leds,
+    /// Buttons
+    pub buttons: Buttons,
+    /// Timer
+    pub timer: Timer,
+    /// uarte interface
+    pub uarte: Uarte,
+}
+
+// ...
+
+pub fn init() -> Result<Board, ()> {
+
+    // ... 
+
+        Ok(Board {
+            leds: Leds {
+                // ...
+            },
+
+            buttons: Buttons {
+                // ...
+            },
+            // 🔼  --- Exercise Button --- 🔼 
+
+            timer: Timer { inner: timer },
+
+            uarte: Uarte { inner: uarte },
+        })
+    } else {
+        Err(())
+    }
+```
+
+</details>
+
+### Step 6: Implementing the `fmt::Write` trait
+
+We can't just write to the `Uarte` instance. A simple write would write from flash memory. This does not work because of EasyDMA. We have to write a function that implements the `fmt::Write` trait. This trait guarantees that the buffer is fully and successfully written on a stack allocated buffer, before it returns. 
```

Comment:
```suggestion
We want to implement the `fmt::Write` trait so that users can call `write!` on our Uarte object

When implementing this, we can't just write to the `Uarte` instance because a simple write of a string literal would try and read the string literal from flash memory. This does not work because the EasyDMA peripheral in the nRF52 series can only access RAM, not flash. 

Instead our implementation must ensure all the strings are copied to a stack allocated buffer and that buffer is passed to the Uarte's `write` method. 
```

Reply:
You could go further here and suggest a performance optimisation to detect which string slices are in DMA-able memory regions and `write` those directly without buffering.